File: android/app/src/main/java/com/rork/porchivo/util/RiskEngine.kt
package com.rork.porchivo.util

import com.rork.porchivo.config.AppConfig
import com.rork.porchivo.model.Shipment
import com.rork.porchivo.model.ShipmentStatus
import java.util.Calendar

/**
 * Porch risk scoring — mirrors RISK_THRESHOLDS.factors in expo/config/app.ts.
 * Score runs 0 (no risk) to 100 (very high risk).
 */
object RiskEngine {

    data class RiskFactor(val label: String, val delta: Int)

    enum class RiskLevel(val label: String) {
        LOW("Low Risk"),
        MEDIUM("Medium Risk"),
        HIGH("High Risk"),
    }

    fun factors(shipments: List<Shipment>): List<RiskFactor> {
        val factors = mutableListOf<RiskFactor>()

        // Mirrors someActiveAlerts (+14) — demo block has 1–2 active alerts
        factors.add(RiskFactor("1–2 theft alerts on your block", +14))

        val active = shipments.filter {
            it.status == ShipmentStatus.OPEN || it.status == ShipmentStatus.ACCEPTED
        }

        if (active.size >= 6) {
            factors.add(RiskFactor("High delivery traffic this week", +12))
        }

        val hasPartner = active.any { it.partnerId != null }
        if (hasPartner) {
            factors.add(RiskFactor("Porch Partner holding your package", -22))
        } else {
            factors.add(RiskFactor("No Porch Partner assigned", +8))
        }

        val lateWindow = active.any { shipment ->
            val cal = Calendar.getInstance().apply { timeInMillis = shipment.deliveryWindowEnd }
            cal.get(Calendar.HOUR_OF_DAY) >= 16
        }
        if (lateWindow) {
            factors.add(RiskFactor("Delivery window after 4 PM", +14))
        } else {
            factors.add(RiskFactor("Daytime delivery window", -4))
        }

        if (active.any { it.notes.isNotBlank() }) {
            factors.add(RiskFactor("Drop instructions added", -4))
        }

        return factors
    }

    fun score(shipments: List<Shipment>): Int {
        val total = AppConfig.RiskThresholds.BASE_SCORE + factors(shipments).sumOf { it.delta }
        return total.coerceIn(0, 100)
    }

    fun level(score: Int): RiskLevel = when {
        score >= AppConfig.RiskThresholds.HIGH -> RiskLevel.HIGH
        score >= AppConfig.RiskThresholds.MEDIUM -> RiskLevel.MEDIUM
        else -> RiskLevel.LOW
    }
}

File: ios/Porchivo/Utilities/RiskEngine.swift
//
//  RiskEngine.swift
//  Porchivo
//
//  Porch risk scoring — mirrors expo RISK_THRESHOLDS.factors and android RiskEngine.kt.
//

import Foundation

enum RiskEngine {
    struct RiskFactor: Identifiable, Equatable {
        let id = UUID()
        let label: String
        let delta: Int
    }

    enum RiskLevel: String {
        case low, medium, high
        var label: String {
            switch self {
            case .low: "Low Risk"
            case .medium: "Medium Risk"
            case .high: "High Risk"
            }
        }
    }

    static func factors(_ shipments: [Shipment]) -> [RiskFactor] {
        var out: [RiskFactor] = []
        out.append(RiskFactor(label: "1–2 theft alerts on your block", delta: 14))
        let active = shipments.filter { $0.status == .open || $0.status == .accepted }
        if active.count >= 6 {
            out.append(RiskFactor(label: "High delivery traffic this week", delta: 12))
        }
        let hasPartner = active.contains { $0.partnerId != nil }
        if hasPartner {
            out.append(RiskFactor(label: "Porch Partner holding your package", delta: -22))
        } else {
            out.append(RiskFactor(label: "No Porch Partner assigned", delta: 8))
        }
        let lateWindow = active.contains { Calendar.current.component(.hour, from: $0.deliveryWindowEnd) >= 16 }
        if lateWindow {
            out.append(RiskFactor(label: "Delivery window after 4 PM", delta: 14))
        } else {
            out.append(RiskFactor(label: "Daytime delivery window", delta: -4))
        }
        if active.contains(where: { !$0.notes.isEmpty }) {
            out.append(RiskFactor(label: "Drop instructions added", delta: -4))
        }
        return out
    }

    static func score(_ shipments: [Shipment]) -> Int {
        let total = AppConfig.RiskThresholds.baseScore + factors(shipments).reduce(0) { $0 + $1.delta }
        return max(0, min(100, total))
    }

    static func level(_ score: Int) -> RiskLevel {
        if score >= AppConfig.RiskThresholds.high { return .high }
        if score >= AppConfig.RiskThresholds.medium { return .medium }
        return .low
    }
}
